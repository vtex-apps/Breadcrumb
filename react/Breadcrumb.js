import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import { injectIntl, intlShape } from 'react-intl'
import { Link } from 'render'

const CSS_CLASSES = {
  BREADCRUMB: 'vtex-breadcrumb pa4 gray',
  LINK: 'vtex-breadcrumb__link dib pv1 link light-primary hover-primary ph2',
}

/**
 * Breadcrumb Component.
 */
class Breadcrumb extends Component {
  getPageHint(categories) {
    switch (categories.length) {
      case 1:
        return 'd'
      case 2:
        return 'c'
      case 3:
        return 'sc'
    }
  }

  getCategories(categories) {
    const categoriesSorted = categories
      .slice()
      .sort((a, b) => a.length - b.length)
    return categoriesSorted.map(category => {
      const categoryStripped = category.replace(/^\//, '').replace(/\/$/, '')
      const categories = categoryStripped.split('/')
      const [categoryKey] = categories.reverse()
      const hint = this.getPageHint(categories)
      return {
        name: categoryKey.toLowerCase(),
        value: `${categoryStripped}/${hint}`,
      }
    })
  }

  render() {
    const { term, categories } = this.props

    if (!categories) {
      return null
    }

    const categoriesList = this.getCategories(categories)
    return (
      <div className={CSS_CLASSES.BREADCRUMB}>
        <Link className={CSS_CLASSES.LINK} page="store">
          {this.props.intl.formatMessage({ id: 'breadcrumb.home' })}
        </Link>
        {categoriesList.map(({ name, value }, i) => (
          <span key={`category-${i}`}>
            /
            <Link className={CSS_CLASSES.LINK} to={`/${value}`}>
              {name}
            </Link>
          </span>
        ))}

        {term && (
          <Fragment>
            /
            <span className="ph2"> {term} </span>
          </Fragment>
        )}
      </div>
    )
  }
}

Breadcrumb.propTypes = {
  /** Intl instance. */
  intl: intlShape.isRequired,
  /** Search term or product slug. */
  term: PropTypes.string,
  /** Product's categories. */
  categories: PropTypes.arrayOf(PropTypes.string),
}

export default injectIntl(Breadcrumb)
