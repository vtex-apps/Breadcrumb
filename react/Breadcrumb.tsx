import PropTypes from 'prop-types'
import React, { Component, Fragment, ReactNode } from 'react'
import { Link } from 'render'

import ArrowIcon from './icons/ArrowIcon'
import HomeIcon from './icons/HomeIcon'

const LINK_CLASS_NAME =
  'vtex-breadcrumb__link dib pv1 link ph2 gray hover-near-black'

interface DefaultProps {
  categories: Array<string>
}
  
interface Props extends DefaultProps {
  term?: string
}

type Category = {
  name: string
  value: string
}

/**
 * Breadcrumb Component.
 */
class Breadcrumb extends Component<Props> {
  public static readonly defaultProps = {
    categories: [],
  }

  public static readonly propTypes = {
    /** Search term or product slug. */
    term: PropTypes.string,
    /** Product's categories. */
    categories: PropTypes.arrayOf(PropTypes.string),
  }

  private get categoriesList(): Array<Category> {
    const { categories } = this.props
    const categoriesSorted = categories
      .slice()
      .sort((a, b) => a.length - b.length)
    return categoriesSorted.map(category => {
      const categoryStripped = category.replace(/^\//, '').replace(/\/$/, '')
      const categories = categoryStripped.split('/')
      const [categoryKey] = categories.reverse()

      return {
        name: categoryKey.toLowerCase(),
        value: categoryStripped,
      }
    })
  }

  public render(): ReactNode {
    const { term, categories } = this.props

    if (categories.length === 0) {
      return null
    }

    return (
      <div className="vtex-breadcrumb dn db-ns pa4 gray">
        <Link className={LINK_CLASS_NAME} page="store/home">
          <HomeIcon />
        </Link>
        {this.categoriesList.map(({ name, value }, i) => (
          <span key={`category-${i}`}>
            <span className="ph2">
              <ArrowIcon />
            </span>
            <Link className={LINK_CLASS_NAME} to={`/${value}`}>
              {name}
            </Link>
          </span>
        ))}

        {term && (
          <Fragment>
            <span className="ph2">
              <ArrowIcon />
            </span>
            <span className="vtex-breadcrumbs__term ph2 near-black">
              {term}
            </span>
          </Fragment>
        )}
      </div>
    )
  }
}

export default Breadcrumb
