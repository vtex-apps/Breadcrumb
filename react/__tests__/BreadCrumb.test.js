/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'

import Breadcrumb from '../Breadcrumb'

describe('<BreadCrumb /> component', () => {
  const renderComponent = customProps => {
    const props = {
      categories: [
        {
          link: '/Eletr么nicos/',
          name: 'Eletr么nicos',
        },
        {
          link: '/Eletr么nicos/Smartphones/',
          name: 'Smartphones :)',
        },
      ],
      ...customProps,
    }
    return render(<Breadcrumb {...props} />)
  }

  it('should be rendered', () => {
    expect(renderComponent().asFragment().firstChild).toHaveClass('container')
  })

  it('should match the snapshot', () => {
    expect(renderComponent({ term: 'term' }).asFragment()).toMatchSnapshot()
  })

  it('should render links', () => {
    const { queryByText, asFragment } = renderComponent()
    expect(asFragment().querySelectorAll('a')).toHaveLength(3)
    expect(queryByText('eletr么nicos')).toBeDefined()
    expect(queryByText('smartphones')).toBeDefined()
  })

  it('should render term', () => {
    const { rerender, queryByText } = renderComponent({
      term: 'term',
    })
    expect(queryByText('term')).toBeDefined()

    rerender(<Breadcrumb />)
    expect(queryByText('term')).toBeNull()
  })
})
