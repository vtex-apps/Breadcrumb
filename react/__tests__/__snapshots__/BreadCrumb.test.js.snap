// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreadCrumb /> component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container dn db-ns pv3"
  >
    <a
      class="link"
    >
      <div
        size="26"
      >
        Icon Home
      </div>
    </a>
    <span
      class="arrow ph2 c-muted-2"
    >
      <div
        orientation="right"
        size="8"
      >
        Icon Caret
      </div>
      <a
        class="link"
      >
        Eletr√¥nicos
      </a>
    </span>
    <span
      class="arrow ph2 c-muted-2"
    >
      <div
        orientation="right"
        size="8"
      >
        Icon Caret
      </div>
      <a
        class="link"
      >
        Smartphones :)
      </a>
    </span>
    <span
      class="arrow ph2 c-muted-2"
    >
      <div
        orientation="right"
        size="8"
      >
        Icon Caret
      </div>
    </span>
    <span
      class="term ph2 c-on-base"
    >
      term
    </span>
  </div>
</DocumentFragment>
`;
